fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/TemplateDruidQuery.java

```diff
@@ -429,6 +439,162 @@ public MetricField getMetricField(String name) {
+     */
+    public TemplateDruidQuery renameMetricField(@NotNull String currentName, @NotNull String newName) {
+        Objects.requireNonNull(currentName, NULL_RENAME_ERROR_MESSAGE);
+        Objects.requireNonNull(newName, NULL_RENAME_ERROR_MESSAGE);
+
+        if (currentName.equals(newName)) {
+            return this;
+        }
+
+        if (Stream.concat(getAggregations().stream(), getPostAggregations().stream())
+                .anyMatch(mf -> Objects.equals(newName, mf.getName()))) {
+            throw new IllegalArgumentException(
+                String.format(RENAME_TO_DUPLICATE_NAME_ERROR_MESSAGE, currentName, newName)
+            );
+        }
+
+        if (getAggregations().stream().anyMatch(agg -> Objects.equals(agg.getName(), currentName))) {
+            return renameAggregation(currentName, newName);
+        }
+
+        if (getPostAggregations().stream().anyMatch(pa -> Objects.equals(pa.getName(), currentName))) {
+            return renamePostAggregation(currentName, newName);
+        }
+
+        throw new IllegalArgumentException(String.format(NO_METRIC_TO_RENAME_FOUND_ERROR_MESSAGE, currentName));
+    }
+
+    /**
+     * Renames the {@link Aggregation} with name {@code currentName}, and renames its output name to {@code newName}.
+     * The field name the target Aggregation references is never directly touched. Additionally, all
+     * {@link PostAggregation}s that reference the renamed Aggregation have their references updated to point to the
+     * renamed Aggregation.
+     * <p>
+     * The behavior when either parameter is null is undefined, as such this method should never be called with null
+     * parameter values.
+     *
+     * @param currentName  The output name of the Aggregation to rename
+     * @param newName  The new output name
+     * @return the TemplateDruidQuery with the renamed Aggregation and all updated PostAggregations
+     * @throws IllegalArgumentException if there is no Aggregation with an output name matching currentName
+     */
+    protected TemplateDruidQuery renameAggregation(String currentName, String newName) {
+        Aggregation renamedAgg = getAggregations().stream()
+                .filter(agg -> agg.getName().equals(currentName))
+                .findFirst()
+                .orElseThrow(IllegalArgumentException::new)
+                .withName(newName);
+
+        Set<Aggregation> newAggs = getAggregations().stream()
+                .filter(agg -> !agg.getName().equals(currentName))
+                .collect(Collectors.toSet());
+        newAggs.add(renamedAgg);
+
+        Set<PostAggregation> newPostAggs = new HashSet<>();
+        for (PostAggregation pa : getPostAggregations()) {
+            // TODO add repoint AggregationReference here.
+            if (pa instanceof WithFields) {
+                newPostAggs.add(renameAggregationReferences(currentName, renamedAgg, (WithFields<?>) pa));
+            } else {
+                newPostAggs.add(pa);
```

Comment:
I don't think this handles the post agg being a Field Access post aggregation

Reply:
Addressed with AggregationReference interface.